feat: draft for 4.3 article (#972)
withastro/astro.build: src/content/blog/astro-430.mdx

publishDate: "February 2, 2024"
authors:
  - erika
  - ema
  - matthew
  - nate
  - bjorn
coverImage: "/src/content/blog/_images/astro-430/post-header-4.3.webp"
socialImage: "/src/content/blog/_images/astro-430/og-image-4.3.webp"
lang: "en"
---

Astro 4.3 is now available! This release

Highlights include:

- **[Add domain support for i18n](#experimental-add-domain-support-for-i18n)**
- **[Add support for `build.format: 'preserve'`](#add-support-for-buildformat-preserve)**
- **[Add `ComponentProps` type utility](#add-componentprops-type-utility)**
- **[Better support for relative images in Markdown](#fix-using-images-in-markdown-without-a-relative-specifier)**

## How to upgrade

To take advantage of the latest features, make sure you're running the latest version of Astro. You can upgrade to Astro 4.2 by running the `@astrojs/upgrade` command:

```sh
npx @astrojs/upgrade
```

or by running the upgrade command for your package manager:

```sh
npm install astro@latest
pnpm upgrade astro --latest
yarn upgrade astro --latest
```

## (Experimental) Add domain support for i18n

> Astro now supports adding a domain to your i18n configuration. This allows you to specify a domain for each locale, which allows you to use different domains or subdomains for different locales. For example, you can now use `example.com` for your English site, and `example.es` for your Spanish site.

Reviewer: comma nit
```suggestion
Astro now supports adding a domain to your i18n configuration. This allows you to specify a domain for each locale, which allows you to use different domains or subdomains for different locales. For example, you can now use `example.com` for your English site and `example.es` for your Spanish site.
Comment
```

To add a domain to your i18n configuration, add the following to your `astro.config.mjs` file:

```diff lang="js"
// astro.config.mjs
import {defineConfig} from "astro/config"

export default defineConfig({
    site: "https://example.com",
    output: "server",
    i18n: {
        defaultLocaLe: 'en',
        locales: ['en', 'es', 'pt_BR', 'pt', 'fr'],
+        domains: {
+            fr: "https://fr.example.com",
+            es: "https://example.es"
+        },
        routing: {
          prefixDefaultLocale: true,
+          strategy: "domains"
        }
    },
+    experimental: {
+        i18nDomains: true
+    },
})
```

> TOOD: Add proper link to final documentation, perhaps some fluff on how this feature requires SSR?

Reviewer: Flagging so we don't miss changing this.

Reviewer 2: @sarah11918 @Yan-Thomas what do you think we should add here?

For more information, check out the [i18n documentation](https://docs.astro.build/reference/configuration#i18n).

## Add support for `build.format: 'preserve'`

This release adds support for a new `build.format` option, called `preserve`. In this mode, Astro will preserve how the filesystem is structured and make sure that is mirrored over to production. For example, using this option:

> - `page.astro` becomes `page.html`
> - `page/index.astro` becomes `page/index.html`

Reviewer 3: ```suggestion
- `about.astro` becomes `about.html`
- `about/index.astro` becomes `about/index.html`
```
*Slight* preference for not using `page` (so that there's no confusion with the `src/pages/` folder!), as I suggested in my review on the feature PR!

What you see is what you get! This feature unlocks better compatibility with certain server softwares who have strict requirements on how files are structured.

For more information on the `build.format` option, check out the [configuration documentation](https://docs.astro.build/en/reference/configuration-reference/#buildformat).

## Add `ComponentProps` type utility

> Astro now includes a new `ComponentProps` type export from `astro/types` to get the props type of an Astro component. For users of other frameworks, this is similar to `React.ComponentProps` or Svelte's `ComponentProps`. Simple enough!

Reviewer 3: ```suggestion
Astro now includes a new `ComponentProps` type export from `astro/types` to get the props type of an Astro component. This is similar to `React.ComponentProps` or Svelte's `ComponentProps`. 

This type export allows you to reference the `Props` accepted by another component, even if that component doesn't export that `Props` type directly.
```

```astro
---
import type { ComponentProps } from 'astro/types';
import { Button } from "./Button.astro";

type myButtonProps = ComponentProps<typeof Button>;
---
```

## Fix using images in Markdown without a relative specifier

> Previously, using images in Markdown without using a relative specifier (such as `./` or `../`) would cause Astro to throw an error. This release fixes that issue, and now allows you to use images in Markdown without a relative specifier, as the Markdown spec intended.

Reviewer: comma nit
```suggestion
Previously, using images in Markdown without using a relative specifier (such as `./` or `../`) would cause Astro to throw an error. This release fixes that issue and now allows you to use images in Markdown without a relative specifier, as the Markdown spec intended.
```

> 
> Previously, using images in Markdown without using a relative specifier (such as `./` or `../`) would cause Astro to throw an error. This release fixes that issue and now allows you to use images in Markdown without a relative specifier, as the Markdown spec intended.
> 
> ```md
> ![A cute dog](dog.jpg)
> <!-- This dog lives in the same folder as my article! -->
> ```

Reviewer 3: ``````suggestion

Previously, using images in Markdown without using a relative specifier (such as `./` or `../`) would cause Astro to throw an error. 

Now, you can use the standard `![](img.png)` syntax in Markdown files for images colocated in the same folder: no relative specifier required!

There is no need to update your project; your existing images will still continue to work. However, you can safely remove any relative specifiers from these Markdown images as they are no longer necessary... as the Markdown spec intended!

```diff
- ![A cute dog](./dog.jpg)
+ ![A cute dog](dog.jpg)
<!-- This dog lives in the same folder as my article! -->
```
``````

---
title: "Astro 4.3"
description: "Astro 4.3 is out now! This release includes support for domain routing in i18n, better support for relative images in Markdown, a new `ComponentProps` type export, and more."
publishDate: "February 2, 2024"
authors:
  - erika
  - ema
  - matthew
  - nate
  - bjorn
coverImage: "/src/content/blog/_images/astro-430/post-header-4.3.webp"
socialImage: "/src/content/blog/_images/astro-430/og-image-4.3.webp"
lang: "en"
---

Astro 4.3 is now available! This release includes a new experimental i18n feature to try out, and improvements to working with your build output, component prop types, Markdown images, and more.

Highlights include:

> - **[Add domain support for i18n](#experimental-add-domain-support-for-i18n)**

Reviewer 4: Makes this bullet match the heading, which seemed to be what other posts do too:

```suggestion
- **[Experimental: Add domain support for i18n](#experimental-add-domain-support-for-i18n)**
```

> - **[Add support for `build.format: 'preserve'`](#add-support-for-buildformat-preserve)**

Reviewer 4: And mirroring that then up here:

```suggestion
- **[More control over your HTML file output](#more-control-over-your-html-file-output)**
```

- **[Add `ComponentProps` type utility](#add-componentprops-type-utility)**
- **[Better support for relative images in Markdown](#fix-using-images-in-markdown-without-a-relative-specifier)**

## How to upgrade

To take advantage of the latest features, make sure you're running the latest version of Astro. You can upgrade to Astro 4.3 by running the `@astrojs/upgrade` command:

```sh
npx @astrojs/upgrade
```

or by running the upgrade command for your package manager:

```sh
npm install astro@latest
pnpm upgrade astro --latest
yarn upgrade astro --latest
```

> ## (Experimental) Add domain support for i18n

Reviewer 4: Colon ftw!

```suggestion
## Experimental: Add domain support for i18n
```

Astro 4.3 adds an experimental `domains` i18n configuration. This allows you to specify different domains or subdomains for different supported locales.

> For example, you can now use `example.com` for your English site, `fr.example.com` for your French site, and `example.es` for your Spanish site.

Reviewer 4: ```suggestion
For example, you could now use `example.com` for your English site, `fr.example.com` for your French site, and `example.es` for your Spanish site.
```

Enable the experimental flag `i18nDomains` and map any or all of your locales to domains using `i18n.domains` in your `astro.config.mjs` file:

```diff lang="js"
// astro.config.mjs
import {defineConfig} from "astro/config"

export default defineConfig({
    site: "https://example.com",
    output: "server", // required, with no prerendered pages
    adapter: node({
      mode: 'standalone',
    }),
    i18n: {
        defaultLocaLe: 'en',
        locales: ['en', 'es', 'pt_BR', 'pt', 'fr'],
+        domains: {
+            fr: "https://fr.example.com",
+            es: "https://example.es"
+        },
        routing: {
          prefixDefaultLocale: true,
        }
    },
+    experimental: {
+        i18nDomains: true
+    },
})
```

Note that this feature requires an entirely server-rendered site with no prerendered pages. Currently, the `@astrojs/node` and `@astrojs/vercel` adapters are supported with more adapter compatibility to come!

See more in our [internationalization documentation](https://docs.astro.build/en/guides/internationalization/#domains-experimental) for more details and limitations on this experimental routing feature.

> ## Add support for `build.format: 'preserve'`

Reviewer 4: `build.format: 'preserve'` is kind of not a feature we support, but a config option that results in support for something. Suggesting this alternative (H/T @sarah11918)

```suggestion
## More control over your HTML file output
```

Rather than introduce a breaking change to `file`,  we added the new `preserve` format which will preserve how the filesystem is structured and make sure that is mirrored over to production:

- `about.astro` becomes `about.html`
- `about/index.astro` becomes `about/index.html`

What you see is what you get! This feature unlocks better compatibility with certain web servers who have strict requirements on how files are structured.

See the [`build.format` configuration options reference](https://docs.astro.build/en/reference/configuration-reference/#buildformat) for more details.

## Add `ComponentProps` type utility

Astro now includes a new `ComponentProps` type export from `astro/types` to get the props type of an Astro component. This is similar to `React.ComponentProps` or Svelte's `ComponentProps`. 

This type export allows you to reference the `Props` accepted by another component, even if that component doesn't export that `Props` type directly.

```astro
---
import type { ComponentProps } from 'astro/types';
import { Button } from "./Button.astro";

> type myButtonProps = ComponentProps<typeof Button>;

Reviewer 4: Nano-nit, types are very often uppercase?

```suggestion
type MyButtonProps = ComponentProps<typeof Button>;
```
